File: csharp/Storage/VectorStore.cs
// Qdrant vector store client using REST API via HttpClient.
// Manages two collections: memory_text and memory_visual.

using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;

namespace AgentMemory.Storage;

/// <summary>Qdrant REST API vector store for memory embeddings.</summary>
public sealed class VectorStore : IDisposable
{
    public const string TextCollection = "memory_text";
    public const string VisualCollection = "memory_visual";

    private readonly HttpClient _client;
    private readonly string _baseUrl;
    private readonly ILogger<VectorStore>? _logger;
    private readonly bool _ownsClient;

    public VectorStore(string? baseUrl = null, HttpClient? client = null, ILogger<VectorStore>? logger = null)
    {
        _baseUrl = (baseUrl ?? "http://localhost:6333").TrimEnd('/');
        _ownsClient = client is null;
        _client = client ?? new HttpClient();
        _logger = logger;
    }

    public async Task InitializeAsync(int textDim = 384, int visualDim = 512)
    {
        await EnsureCollectionAsync(TextCollection, textDim);
        await EnsureCollectionAsync(VisualCollection, visualDim);
    }

    public void Dispose()
    {
        if (_ownsClient)
            _client.Dispose();
    }

    private async Task EnsureCollectionAsync(string name, int dim)
    {
        try
        {
            var resp = await _client.GetAsync($"{_baseUrl}/collections/{name}");
            if (resp.IsSuccessStatusCode)
                return;
        }
        catch { /* collection doesn't exist */ }

        var body = JsonSerializer.Serialize(new
        {
            vectors = new { size = dim, distance = "Cosine" }
        });

        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        await _client.PutAsync($"{_baseUrl}/collections/{name}", content);
    }

    // ── Text embeddings ──

    /// <summary>Insert or update a text embedding. Returns the point ID.</summary>
    public async Task<string> UpsertTextVectorAsync(
        string memoryId, List<double> vector,
        string tier = "hot", double valence = 0.0, double arousal = 0.0,
        string sessionId = "", string createdAt = "")
    {
        var pointId = Guid.NewGuid().ToString();

        var body = JsonSerializer.Serialize(new
        {
            points = new[]
            {
                new
                {
                    id = pointId,
                    vector = vector,
                    payload = new Dictionary<string, object>
                    {
                        ["memory_id"] = memoryId,
                        ["tier"] = tier,
                        ["valence"] = valence,
                        ["arousal"] = arousal,
                        ["session_id"] = sessionId,
                        ["created_at"] = createdAt,
                    }
                }
            }
        });

        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        await _client.PutAsync($"{_baseUrl}/collections/{TextCollection}/points", content);

        return pointId;
    }

    /// <summary>Search for nearest text embeddings.</summary>
    public async Task<List<VectorSearchResult>> SearchTextAsync(
        List<double> queryVector, int limit = 5, string? tierFilter = null)
    {
        var request = new Dictionary<string, object>
        {
            ["vector"] = queryVector,
            ["limit"] = limit,
            ["with_payload"] = true,
        };

        if (tierFilter is not null)
        {
            request["filter"] = new
            {
                must = new[]
                {
                    new { key = "tier", match = new { value = tierFilter } }
                }
            };
        }

        var body = JsonSerializer.Serialize(request);
        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        var resp = await _client.PostAsync($"{_baseUrl}/collections/{TextCollection}/points/search", content);
        var json = await resp.Content.ReadAsStringAsync();

        return ParseSearchResults(json);
    }

    // ── Visual embeddings ──

    /// <summary>Insert or update a visual (CLIP) embedding. Returns the point ID.</summary>
    public async Task<string> UpsertVisualVectorAsync(
        string memoryId, List<double> vector,
        string sessionId = "", string createdAt = "")
    {
        var pointId = Guid.NewGuid().ToString();

        var body = JsonSerializer.Serialize(new
        {
            points = new[]
            {
                new
                {
                    id = pointId,
                    vector = vector,
                    payload = new Dictionary<string, object>
                    {
                        ["memory_id"] = memoryId,
                        ["session_id"] = sessionId,
                        ["created_at"] = createdAt,
                    }
                }
            }
        });

        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        await _client.PutAsync($"{_baseUrl}/collections/{VisualCollection}/points", content);

        return pointId;
    }

    /// <summary>Search for nearest visual embeddings.</summary>
    public async Task<List<VectorSearchResult>> SearchVisualAsync(
        List<double> queryVector, int limit = 5)
    {
        var body = JsonSerializer.Serialize(new
        {
            vector = queryVector,
            limit,
            with_payload = true,
        });

        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        var resp = await _client.PostAsync($"{_baseUrl}/collections/{VisualCollection}/points/search", content);
        var json = await resp.Content.ReadAsStringAsync();

        return ParseSearchResults(json);
    }

    /// <summary>
    /// Compute cosine similarity between two points in the text collection.
    /// Used by dream explorer (A3) for cross-session similarity checks.
    /// </summary>
    public async Task<double?> SimilarityAsync(string pointIdA, string pointIdB)
    {
        try
        {
            var body = JsonSerializer.Serialize(new
            {
                ids = new[] { pointIdA, pointIdB },
                with_vector = true,
            });

            var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
            var resp = await _client.PostAsync(
                $"{_baseUrl}/collections/{TextCollection}/points", content);
            var json = await resp.Content.ReadAsStringAsync();

            using var doc = JsonDocument.Parse(json);
            var result = doc.RootElement.GetProperty("result");

            if (result.GetArrayLength() < 2)
                return null;

            var vecA = result[0].GetProperty("vector").EnumerateArray().Select(e => e.GetDouble()).ToArray();
            var vecB = result[1].GetProperty("vector").EnumerateArray().Select(e => e.GetDouble()).ToArray();

            return CosineSimilarity(vecA, vecB);
        }
        catch
        {
            return null;
        }
    }

    /// <summary>Delete all points for a given memory_id from a collection.</summary>
    public async Task DeletePointAsync(string collection, string memoryId)
    {
        var body = JsonSerializer.Serialize(new
        {
            filter = new
            {
                must = new[]
                {
                    new { key = "memory_id", match = new { value = memoryId } }
                }
            }
        });

        var content = new StringContent(body, System.Text.Encoding.UTF8, "application/json");
        await _client.PostAsync($"{_baseUrl}/collections/{collection}/points/delete", content);
    }

    // ── Helpers ──

    private static double CosineSimilarity(double[] a, double[] b)
    {
        double dot = 0, normA = 0, normB = 0;
        for (var i = 0; i < a.Length && i < b.Length; i++)
        {
            dot += a[i] * b[i];
            normA += a[i] * a[i];
            normB += b[i] * b[i];
        }

        var norm = Math.Sqrt(normA) * Math.Sqrt(normB);
        return norm > 0 ? dot / norm : 0.0;
    }

    private static List<VectorSearchResult> ParseSearchResults(string json)
    {
        var results = new List<VectorSearchResult>();
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (!doc.RootElement.TryGetProperty("result", out var resultArray))
                return results;

            foreach (var item in resultArray.EnumerateArray())
            {
                var payload = item.GetProperty("payload");
                var result = new VectorSearchResult
                {
                    MemoryId = payload.TryGetProperty("memory_id", out var mid) ? mid.GetString() ?? "" : "",
                    Score = item.TryGetProperty("score", out var s) ? s.GetDouble() : 0.0,
                    Tier = payload.TryGetProperty("tier", out var t) ? t.GetString() ?? "hot" : "hot",
                    Valence = payload.TryGetProperty("valence", out var v) ? v.GetDouble() : 0.0,
                    Arousal = payload.TryGetProperty("arousal", out var a) ? a.GetDouble() : 0.0,
                };
                results.Add(result);
            }
        }
        catch { /* parse failure — return empty */ }

        return results;
    }
}

/// <summary>Search result from the vector store.</summary>
public sealed class VectorSearchResult
{
    public string MemoryId { get; set; } = "";
    public double Score { get; set; }
    public string Tier { get; set; } = "hot";
    public double Valence { get; set; }
    public double Arousal { get; set; }
}

File: agent_memory/storage/vector_store.py
"""Qdrant vector store for semantic similarity search.

Runs in embedded (local) mode — no server required. Manages two collections:
  - memory_text: sentence-transformer embeddings of memory content
  - memory_visual: CLIP embeddings of scene descriptions
"""

from __future__ import annotations

import logging
import uuid
from pathlib import Path
from typing import TYPE_CHECKING

from agent_memory.config import VECTOR_DIR

if TYPE_CHECKING:
    from qdrant_client import QdrantClient

logger = logging.getLogger(__name__)

TEXT_COLLECTION = "memory_text"
VISUAL_COLLECTION = "memory_visual"


class VectorStore:
    """Qdrant-backed vector store for memory embeddings.

    ``qdrant_client`` is imported lazily so this module is importable without
    the ``vectors`` extra. Profiles that retrieve only via grep/keyword (e.g.
    the lite/edge profile) need not install Qdrant.
    """

    def __init__(self, vector_dir: Path | None = None) -> None:
        self.vector_dir = vector_dir or VECTOR_DIR
        self._client: QdrantClient | None = None

    def initialize(self, text_dim: int = 384, visual_dim: int = 512) -> None:
        """Initialize Qdrant in embedded mode and ensure collections exist."""
        try:
            from qdrant_client import QdrantClient
        except ImportError as exc:  # pragma: no cover - dependency guard
            raise ImportError(
                "VectorStore requires the 'vectors' extra. "
                "Install with: pip install agent-memory[vectors]"
            ) from exc
        self.vector_dir.mkdir(parents=True, exist_ok=True)
        self._client = QdrantClient(path=str(self.vector_dir))
        self._ensure_collection(TEXT_COLLECTION, text_dim)
        self._ensure_collection(VISUAL_COLLECTION, visual_dim)

    def close(self) -> None:
        if self._client:
            self._client.close()
            self._client = None

    @property
    def client(self) -> QdrantClient:
        assert self._client is not None, "VectorStore not initialized — call initialize() first"
        return self._client

    def _ensure_collection(self, name: str, dim: int) -> None:
        from qdrant_client.models import Distance, VectorParams

        collections = [c.name for c in self.client.get_collections().collections]
        if name in collections:
            # Guard against reopening a data directory with an embedder whose
            # output dimension differs from the existing collection. Qdrant would
            # otherwise silently reject upserts/searches and semantic retrieval
            # would degrade with only swallowed errors. Fail clearly instead.
            try:
                existing_dim = self.client.get_collection(name).config.params.vectors.size
            except Exception:
                existing_dim = None
            if existing_dim is not None and existing_dim != dim:
                raise ValueError(
                    f"Vector collection '{name}' was created with dimension "
                    f"{existing_dim}, but the configured embedder produces dimension "
                    f"{dim}. The embedding provider/model changed for an existing "
                    f"data directory. Use a fresh data directory, or an embedder "
                    f"whose dimension is {existing_dim}."
                )
            return
        self.client.create_collection(
            collection_name=name,
            vectors_config=VectorParams(size=dim, distance=Distance.COSINE),
        )

    # ── Text embeddings ──

    def upsert_text_vector(
        self, memory_id: str, vector: list[float],
        tier: str = "hot", valence: float = 0.0, arousal: float = 0.0,
        session_id: str = "", created_at: str = "", namespace: str = "default",
    ) -> str:
        """Insert or update a text embedding. Returns the point ID."""
        from qdrant_client.models import PointStruct

        point_id = str(uuid.uuid4())
        self.client.upsert(
            collection_name=TEXT_COLLECTION,
            points=[
                PointStruct(
                    id=point_id,
                    vector=vector,
                    payload={
                        "memory_id": memory_id,
                        "tier": tier,
                        "valence": valence,
                        "arousal": arousal,
                        "session_id": session_id,
                        "created_at": created_at,
                        "namespace": namespace,
                    },
                )
            ],
        )
        return point_id

    def search_text(
        self, query_vector: list[float], limit: int = 5,
        tier_filter: str | None = None, namespace: str | None = None,
    ) -> list[dict]:
        """Search for nearest text embeddings.

        Returns list of dicts: {memory_id, score, tier, valence, arousal}.
        Results are restricted to ``namespace`` when provided.
        """
        from qdrant_client.models import FieldCondition, Filter, MatchValue

        must = []
        if tier_filter:
            must.append(FieldCondition(key="tier", match=MatchValue(value=tier_filter)))
        if namespace is not None:
            must.append(FieldCondition(key="namespace", match=MatchValue(value=namespace)))
        search_filter = Filter(must=must) if must else None

        results = self.client.query_points(
            collection_name=TEXT_COLLECTION,
            query=query_vector,
            limit=limit,
            query_filter=search_filter,
        )
        return [
            {
                "memory_id": r.payload["memory_id"],
                "score": r.score,
                "tier": r.payload.get("tier", "hot"),
                "valence": r.payload.get("valence", 0.0),
                "arousal": r.payload.get("arousal", 0.0),
            }
            for r in results.points
        ]

    # ── Visual embeddings ──

    def upsert_visual_vector(
        self, memory_id: str, vector: list[float],
        session_id: str = "", created_at: str = "", namespace: str = "default",
    ) -> str:
        """Insert or update a visual (CLIP) embedding. Returns the point ID."""
        from qdrant_client.models import PointStruct

        point_id = str(uuid.uuid4())
        self.client.upsert(
            collection_name=VISUAL_COLLECTION,
            points=[
                PointStruct(
                    id=point_id,
                    vector=vector,
                    payload={
                        "memory_id": memory_id,
                        "session_id": session_id,
                        "created_at": created_at,
                        "namespace": namespace,
                    },
                )
            ],
        )
        return point_id

    def search_visual(
        self, query_vector: list[float], limit: int = 5,
        namespace: str | None = None,
    ) -> list[dict]:
        """Search for nearest visual embeddings.

        Returns list of dicts: {memory_id, score}. Restricted to ``namespace``
        when provided.
        """
        from qdrant_client.models import FieldCondition, Filter, MatchValue

        search_filter = None
        if namespace is not None:
            search_filter = Filter(
                must=[FieldCondition(key="namespace", match=MatchValue(value=namespace))]
            )
        results = self.client.query_points(
            collection_name=VISUAL_COLLECTION,
            query=query_vector,
            limit=limit,
            query_filter=search_filter,
        )
        return [
            {"memory_id": r.payload["memory_id"], "score": r.score}
            for r in results.points
        ]

    def similarity(self, point_id_a: str, point_id_b: str) -> float | None:
        """Compute cosine similarity between two points in the text collection.

        Returns the similarity score or None if either point is not found.
        Used by dream explorer (A3) for cross-session similarity checks.
        """
        try:
            points = self.client.retrieve(
                collection_name=TEXT_COLLECTION,
                ids=[point_id_a, point_id_b],
                with_vectors=True,
            )
            if len(points) < 2:
                return None
            import numpy as np
            a = np.array(points[0].vector)
            b = np.array(points[1].vector)
            dot = np.dot(a, b)
            norm = np.linalg.norm(a) * np.linalg.norm(b)
            return float(dot / norm) if norm > 0 else 0.0
        except Exception:
            return None

    def delete_point(self, collection: str, memory_id: str) -> None:
        """Delete all points for a given memory_id from a collection."""
        from qdrant_client.models import FieldCondition, Filter, MatchValue

        self.client.delete(
            collection_name=collection,
            points_selector=Filter(
                must=[FieldCondition(key="memory_id", match=MatchValue(value=memory_id))]
            ),
        )
